### ISSUE TITLE
Data type issue with BioSemi ActiTwo

### ISSUE THREAD

**Issue by marcpabst** [mne-tools/mne-lsl#378]
I've tested `mne-lsl` with [https://github.com/labstreaminglayer/App-BioSemi](https://github.com/labstreaminglayer/App-BioSemi) and most things seem to work, but when I obtain a new chunk of samples using `get_data()`, the resulting numpy array consists of `floats` that are extremely negative (e.g., scaled from -5431.59 to -772.0). I suspect there's a mismatch between data types somewhere and bytes are just wrongly interpreted. Other LSL viewer software works okay.

**Comment by mscheltienne** [comment]
It's entirely possible that the `App-Biosemi` is streaming those values. On the `mne-lsl` side, the array you are getting from `StreamLSL.get_data()` is an aggregation of consecutive `StreamOutlet.pull_chunk` operations, which comes down to: https://github.com/mne-tools/mne-lsl/blob/ee8d56ee68bf7c12dad9901a0559df86f146e688/src/mne_lsl/lsl/stream_inlet.py#L396-L398

The datatype used to interpret the bytes is provided by the LSL stream as well, i.e. by the `App-Biosemi`. 

Maybe the viewers you used are applying filters or other processing on the stream, which are absent from:

```python
import time

from mne_lsl.stream import StreamLSL

stream = StreamLSL(bufsize=2).connect()
time.sleep(2)
data = stream.get_data()
```

You can of course apply such filters on the stream as well, e.g. with the [filter](https://mne.tools/mne-lsl/stable/generated/api/mne_lsl.stream.StreamLSL.html#mne_lsl.stream.StreamLSL.filters) method.

----------------

Overall, `-5431.59` to `-772.0` doesn't strike me as strange values for a raw un-processed EEG dataset which usually suffer from DC offset and drifts. For instance, I opened a raw recording from an ANT Neuro eego amplifier with 64 channels gel-caps. In 10s, the recording shows values between `-0.0655369` and `0.00927377` Volts. Moreover, the `eego2lsl` app (equivalent to `App-Biosemi`) streams data in uV, thus the values in the stream would be oscillating between `-65 5536.98` and `9 273.77`.

**Comment by marcpabst** [comment]
> Overall, -5431.59 to -772.0 doesn't strike me as strange values for a raw un-processed EEG dataset which usually suffer from DC offset and drifts. For instance, I opened a raw recording from an ANT Neuro eego amplifier with 64 channels gel-caps. In 10s, the recording shows values between -0.0655369 and 0.00927377 Volts. Moreover, the eego2lsl app (equivalent to App-Biosemi) streams data in uV, thus the values in the stream for this recording would be oscillating between -65 536.98 and 9 273.77.

That's a good point - however, I should have mentioned that the values also seem sign-swapped, i.e. I have a photosensor connected to the amplifier and the signal goes down when I increase the light input to the sensor. Again, other viewers don't show this problem.

But you're comments have prompted me to have closer look once I'm back in the lab!

**Comment by mscheltienne** [comment]
For a photosensor, or for EEG caps/bipolars channels as well, the "direction of change" depends on the wiring, i.e. what is connected to the `+` and what is connected to the `-`. For instance, in our facility we have both: 
- bipolar photodiodes (homemade, with touchproof connectors) which we can sign-swap by swapping the touchproof connectors
- photodiodes to connect to a BrainProduct StimTrak on a 5-pin connector. This 5-pin connector also has a `+` and `-` pin (indicated in the pin layout of the manual); and if I would exchange the 2 wires connected on those pins, I would get a sign-swapped signal.

Overall, a sign-swapped signal just means that you have in your mind a convention/preference for the "direction of change"; and that the wiring of your measuring component has a `+` and `-` swapped compared to this convention/preference. 

The fact that the photodiode actually shows a variation correlated to the light change (even if sign-swapped) is actually very reassuring and sign of a working system.

**Comment by marcpabst** [comment]
You’re absolutely right - it’s not really an issue as long as the relative values remain plausible, which they do. However, I’d still like to understand why both the BrainVision LSL Viewer and BioSemi’s own (non-LSL) viewer display the correct orientation for the signal. Could there be something in the LSL stream that other LSL viewers recognise but is somehow missed by mne-lsl?

**Comment by marcpabst** [comment]
I think you’re right, although it feels somewhat counterintuitive that BrainProducts would implement specific handling for BioSemi channels. I’ll conduct a full test with actual EEG data on Monday and share the results afterwards.

**Comment by mscheltienne** [comment]
Closing as this seems resolved or stalled, feel free to re-open.
